Feat: 회원 API 추가 및 변경, Swagger 2.9.2 연동, Redis 적용

seonpilKim opened the pull request:
### 회원 API 추가 및 변경(Redis 적용)
- [x] 아이디/비밀번호 찾기
- [x] 아이디/이메일/닉네임 중복 체크
- [x] 이메일 인증 코드 발송/인증 코드 검증
- `Gmail SMTP`, `spring-boot-starter-mail`, `Redis cache`를 이용한 메일 전송 로직 추가
  - **회원가입, 아이디/비밀번호 찾기 API**에 `이메일 인증` 로직 포함 
  - 도메인 생성 후 `Gmail SMTP` -> `Mailgun`으로 업데이트 예정
- [x] 비즈니스 예외 및 핸들러, 예외/응답 코드 추가
### 이메일 인증 로직
#### 회원가입
1. Client가 회원가입 페이지로 이동 -> Front에서 `token(숫자, 문자로 구성된 16자리 난수)` 생성
2. Client가 이메일 입력 -> Front에서 텍스트 변경 감지 -> Server에 `이메일 중복 체크 API` ajax 요청
3. Client가 이메일 인증 버튼 클릭 -> Server에 `이메일 인증 코드 발송 API` ajax 요청
4. Server에서 Client에게 인증코드 메일 전송 -> Server에서 `인증 코드와 만료 일자(3분)`를 Front에 응답
    - Front에서 생성한 token을 key로 Redis cache에 value({인증 코드, 만료 일자})와 함께 저장된 상태임
        > (token, {인증 코드, 만료 일자})
    - 만약, 회원이 서버 문제 등으로 이메일을 받지 못하여 재전송 요청을 할 경우, 마지막 요청에 대한 인증 코드만 유효해야 함
        - Redis cache에 token이 key인 value에 새로운 인증 코드와 만료 일자를 덮어씌우는 방식
5. Front에서 인증 코드 일치 여부를 확인 후, Server에 `인증 코드 검증 API` 요청 -> Server에서 `인증서 토큰, 만료 일자(30분)`를 Front에 응답
    > Client가 이메일 인증 완료 이후 30분 이상 멈춰있는 경우, 인증서 토큰이 만료되어 재발급 받아야 회원가입 가능
6. Client가 나머지 입력 칸을 모두 작성하고, 등록 -> Server에 `회원가입 API` 요청 -> Server에서 인증서 토큰 검증 -> 인증서가 유효하면, 회원가입 성공 처리 + 인증서 캐시에서 삭제
    > 인증서 토큰은 이메일 인증이 성공하였다는 내용의 데이터와 함께 Redis cache에 (key, value)로 저장되어 있습니다.
#### 아이디/비밀번호 찾기
1. Client가 회원가입 페이지로 이동 -> Front에서 `token(숫자, 문자로 구성된 16자리 난수)` 생성
2. Client가 이메일 입력 -> Front에서 텍스트 변경 감지 -> Server에 `이메일 중복 체크 API` ajax 요청
3. Client가 이메일 인증 버튼 클릭 -> Server에 `이메일 인증 코드 발송 API` ajax 요청
4. Server에서 Client에게 인증코드 메일 전송 -> Server에서 `인증 코드와 만료 일자(3분)`를 Front에 응답
5. Front에서 인증 코드 일치 여부를 확인 후, Server에 `인증 코드 검증 API` 요청 -> Server에서 `인증서 토큰, 만료 일자(30분)`를 Front에 응답
6. Server에 `아이디/비밀번호 찾기 API` 요청
    - **아이디 찾기**: 인증서 검증 후, 해당 회원의 아이디를 Front에 응답
    - **비밀번호 찾기**: 인증서 토큰과 함께 아이디, 새로운 비밀번호, 비밀번호 확인을 서버에 전달 -> 인증서 검증 후 해당 요청 처리
### Swagger 2.9.2 연동
- [Swagger UI 소개(바로가기)](https://lilac-freeze-4ae.notion.site/Swagger-588c4a1f00044eda80491a28459ccae6)
- Swagger UI HTML(서버 On): http://localhost:8080/swagger-ui.html
- Swagger UI JSON(서버 On): http://localhost:8080/v2/api-docs
- Swagger Hub: https://app.swaggerhub.com/apis/seonpilKim/Gradu/1.0.0#/
___
> p.s) Redis를 이용한 JWT refresh token 방식은 아직 미구현 상태입니다! -> 금방 구현해서 PR 올리겠습니다

Resolve : #7

seonpilKim committed:
Feat: 회원 Entity 업데이트

* [x] userid field 추가

Resolve: #7

seonpilKim committed:
Feat: 회원 API 추가 및 Swagger 3.0 연동

* [x] `Gmail SMTP`, `spring-boot-starter-mail`을 이용한 메일 전송 로직 구현
  - MailService 인터페이스, MailServiceGmailSTMP 구현체 생성
  - application.yml에 관련 설정 추가
  - 추후 도메인 결정 및 생성 후 `MailGun API` 도입 예정
* [x] 회원 이메일 인증 API 추가
  - MailTemplate, MailDTO, MailCertificationResponse 추가
  - MemberController에 `certifyEmail` API 추가
    - 해당 이메일로 메일 전송
    - 6자리 난수 코드와 만료일을 Response
    - Request Header의 Authorization에서 JWT를 받아와서 닉네임을 반환하는 함수 추가 -> `getUsernameFromJwt`
* [x] 회원 아이디 & 닉네임 중복 체크 API 추가
* [x] Spring Boot와 Swagger 3.0 연동
  - `implementation "io.springfox:springfox-boot-starter:3.0.0"` 의존성 추가
  - `SwaggerConfig` 추가
  - Controller, DTO에 어노테이션 적용

Resolve: #7

seonpilKim committed:
Feat: Response, Security 관련 업데이트

* [x] ResultCode, ErrorCode, Exception, GlobalExceptionHandler 업데이트
  - 401 Unauthorized (by Security) Response -> ErrorResponse 반환하도록 Security 로직 수정(`JwtAuthenticationEntryPoint`)

Resolve: #7

seonpilKim committed:
Feat: Swagger 2.9.2 버전 다운, Redis 설정 추가

- Swagger 3.0은 버그가 많고 불안정하므로, 가장 사용자가 많고, 안정적인 2.9.2 버전으로 다운그레이드
- Redis 설정 추가
  - `implementation 'org.springframework.boot:spring-boot-starter-data-redis'`

seonpilKim committed:
Feat: API 추가 및 변경, Redis 적용, DTO 이름 변경

- Redis 적용 API
  - 회원 이메일 인증 API
  - 회원 아이디/비밀번호 찾기 API
  - 회원가입 API
- DTO 이름 변경
  - xxxRequest, xxxResponse로 명확하게 구분

> Redis 적용한 API 로직은 정기 회의 끝나고 업데이트 하겠습니다.

Resolve: #7

seonpilKim committed:
Feat: 예외, 예외 처리 handler, 예외&응답 코드 추가

Resolve: #7

seonpilKim committed:
Refactor: 인증서 토큰 새로 발급하는 방식으로 변경

- Front에서 전달한 토큰을 그대로 사용하는 방식 -> 새로 발급하는 방식

See Also: #7

NohYeaJin reviewed (approved):
고생하셨습니다! 코드가 많아서 자세한 로직은 저도 받아서 실행해봐야할 것 같은데 현재까지 읽힌 코드로는 문제 없는 것 같습니다

1. DTO 파일명 바꾸신 부분 저번 이슈 때문에 명시적으로 바꾸신건지(저도 통일성을 위해 바꿔야할 것 같아서) 여쭤봐도 될까요?

2. 그리고 implementation 'com.querydsl:querydsl-jpa' (build.gradle) 이 부분 추가해주실 수 있을까요? => 헤로쿠가 먼저 빌드해보고 에러 리턴하는 구조이다 보니까 의존성 부분에 넣어놔야 에러가 안떠서 그렇습니다!

3.  ddl-auto: update => 이건 제가 서버 배포할 때 DB 서버 오류 때문에 변경해놨던 사항인데 지금 필요한 건 아니지만 혹시 나중에 충돌날까봐 써놓겠습니다 create로 설정해놓으셔도 제가 update로 바꿔놓으니까 딱히 문제될 부분은 없습니다!

seonpilKim commented:
@NohYeaJin 

1. 넵 xxxRequest, xxxResponse로 구분하는게 직관적으로 혼동이 적을 것 같아서 바꾸었습니다!
2. 추가해서 commit 하겠습니다!
3. 음 배포용 application.yml을 따로 만들어서 관리할까요?
    > 실제 배포할때는 application.yml에 들어가는 민감한 정보들은 따로 환경변수 이용해서 넣어줘야 한다는 얘기를 들어본 것 같아요!

seonpilKim committed:
Docs: @ApiModel value -> description 변경

- value에 한글이 들어가면, Swagger Hub에서 문서화할 때 오류 발생
- description에 한글이 존재하는 것은 문제가 안됨

See Also: #7

seonpilKim committed:
Fix: 인증 코드 검증 -> 인증 코드 만료 예외 처리, Querydsl 의존성 추가

- `verifyCertificationCode`: 인증 코드 만료 에외 처리 추가
- ErrorCode, Exception, Exception Handler 추가
- Querydsl 의존성: `implementation 'com.querydsl:querydsl-jpa'` 추가

Resolve: #7

NohYeaJin commented:
@seonpilKim 
아 그렇네요...! 민감정보는 아예 gitignore에 yml 파일을 추가해서 github에도 안올리도록 해야하는 것 같아요
헤로쿠도 git commit 단위로 서버 배포가 이루어지니까 gitignore에 민감정보를 담은 파일들 다 넣어놓으면 될 것 같습니다!

seonpilKim commented:
@NohYeaJin 

넵 좋습니다! 그럼 `application-dev.yml`, `application-test.yml`, `application-prod.yml` 이렇게 나누면 어떨까요?
(dev는 개발환경, test는 헤로쿠 배포 테스트환경, prod는 실제 운영 배포환경)

괜찮으시면 현재 application.yml은 제거하고, gitignore에 application.yml 관련 추가해서 commit 하겠습니다!
음 그리고 yml파일은 따로 노션에 저장해서 복사해서 사용하면 될 것 같은데 어떻게 생각하시나요?

seonpilKim committed:
Etc: gitignore에 application.yml 추가
